src/us/kbase/workspace/test/database/provenance/TitleTest.java

```diff
@@ -0,0 +1,225 @@
+
+import org.junit.Test;
+
+import nl.jqno.equalsverifier.EqualsVerifier;
+import us.kbase.common.test.TestCommon;
+import static us.kbase.common.test.TestCommon.ES;
+import static us.kbase.workspace.test.database.provenance.ProvenanceTestCommon.WHITESPACE_STRINGS_WITH_NULL;
+
+import us.kbase.workspace.database.provenance.Title;
+import us.kbase.workspace.database.provenance.Title.TitleType;
+
+public class TitleTest {
+	static final String INCORRECT_TITLE = "incorrect title string";
+	static final String INCORRECT_TYPE = "incorrect title type";
+	static final String INCORRECT_LANG = "incorrect title language";
+	static final String EXP_EXC = "expected exception";
+
+	static final String TITLE_STRING = "A Series of Unfortunate Elephants";
+	static final String TITLE_STRING_UNTRIMMED = "\t\t\f\t\n\rA Series of Unfortunate Elephants\n\n\n   ";
+
+	static final String TYPE_STRING = "translated_title";
+	static final String TYPE_STRING_UNTRIMMED = "  \n translated_title   \t\n  ";
+
+	static final String LANG_STRING = "de";
+	static final String LANG_STRING_UNTRIMMED = "\nde\n";
+
+	static final TitleType DEFAULT_TITLE_TYPE = TitleType.TITLE;
+
+	@Test
+	public void equals() throws Exception {
+		EqualsVerifier.forClass(Title.class).usingGetClass().verify();
+	}
+
+	private void assertTitleFields(final Title title, TitleType titleType, String titleLanguage) {
+		assertThat(INCORRECT_TITLE, title.getTitleString(), is(TITLE_STRING));
+		assertThat(INCORRECT_TYPE, title.getTitleType(), is(
+				titleType == null ? TitleType.TITLE : titleType));
+		assertThat(INCORRECT_LANG, title.getTitleLanguage(), is(
+				titleLanguage == null ? ES : opt(titleLanguage)));
+	}
+
+	private void assertTitleFields(final Title title) {
+		assertTitleFields(title, null, null);
+	}
+
+	@Test
+	public void buildMinimal() throws Exception {
+		final Title title = Title.getBuilder(TITLE_STRING).build();
+		assertTitleFields(title);
+	}
+
+	@Test
+	public void buildMaximal() throws Exception {
+		final Title title = Title.getBuilder(TITLE_STRING)
+				.withTitleType(TYPE_STRING)
+				.withTitleLanguage(LANG_STRING).build();
+		assertTitleFields(title, TitleType.TRANSLATED_TITLE, LANG_STRING);
+	}
+
+	@Test
+	public void buildMaximalWithTitleTypeClass() throws Exception {
+		final Title title = Title.getBuilder(TITLE_STRING)
+				.withTitleType(TitleType.OTHER)
+				.withTitleLanguage(LANG_STRING).build();
+		assertTitleFields(title, TitleType.OTHER, LANG_STRING);
+	}
+
+	@Test
+	public void buildTrimAllFields() throws Exception {
+		final Title title = Title.getBuilder(TITLE_STRING_UNTRIMMED)
+				.withTitleType(TYPE_STRING_UNTRIMMED)
+				.withTitleLanguage(LANG_STRING_UNTRIMMED).build();
+		assertTitleFields(title, TitleType.TRANSLATED_TITLE, LANG_STRING);
+	}
+
+	@Test
+	public void buildTransformTitleLanguage() throws Exception {
+		// these tags are valid BCP-47 syntax, even if the values themselves are
+		// nonsense.
+		final String[][] langPairs = {
```

Comment:
It started off as an array but turned into an array of arrays after I decided to add in a formatting element (e.g. input: `ZH_HANT_CN` => output: `zh-Hant-CN`).

I also dislike the standard Map because there's no graceful way to initialise it with values in Java 8. Even the ImmutableMap from Guava maxes out at five key-value pairs (which is sensible, given the syntax, but why not use a different syntax that's less prone to confusion?). Grumble grumble.

Reply:
I do this: https://github.com/kbase/workspace_deluxe/blob/3d716809d838fd46d0d3170cfea59a647bf0a1b6/src/us/kbase/workspace/test/kbase/KBaseWorkspaceConfigTest.java#L441-L449

... which is still clunkier than a literal, but better than Guava or regular maps IMO

Reply:
also java 9 has https://docs.oracle.com/javase/9/docs/api/java/util/Map.html#of-K-V-K-V-K-V-K-V-K-V-K-V-K-V-K-V-K-V-K-V-

Unfortunately getting off 8 is going to be a bear. I've started gradelizing stuff in my spare time to prep for that but it's taking  a while

Reply:
Yes, I suspected that would be the case! At least Java maintains LTS for old versions so keeping a codebase at v8 is a viable (if not always optimal) option.